RFP-01.md

```diff
@@ -0,0 +1,169 @@
+     * @param cowAmt Amount of COW to add as collateral
+     * @param backendUri The URI to send the batches to the solver
+     * @dev Emits an event `SubPoolCreated` when creating a sub-pool
+     */
+    function create(IERC20 token, uint256 amt, uint256 cowAmt, string calldata backendUri) external;
+
+    // --- cow dao bonding pool authed functions ---
+
+    /**
+     * Set the delay for a sub-pool to exit
+     * @dev Emits an event when setting a delay
+     * @param delay The delay in seconds
+     */
+    function setExitDelay(uint256 delay) external auth;
+    /**
+     * Allow `token` to be used as collateral for the sub-pool.
+     * @dev Emits an event `AllowCollateral` when allowing/updating a token
+     */
+    function allowCollateral(IERC20 token, uint256 min) external auth;
+    /**
+     * Revoke `token` as a collateral for sub-pools
+     * @dev Emits an event `RevokeCollateral` when revoking a token
+     */
+    function revokeCollateral(IERC20 token) external auth;
+    /**
+     * Fine a sub-pool
+     * @param pool Which sub-pool to fine
+     * @param tokens The tokens to fine (sending to `msg.sender`)
+     * @param amts The amount of tokens to fine
+     * @dev Emits an event `Fine` when fining a sub-pool
+     */
+    function fine(SubPool pool, IERC20[] calldata tokens, uint256[] calldata amts) external auth;
+
+    // --- SubPool authed functions ---
+
+    /**
+     * Indicate a sub-pool's intent to quit the batch auction
+     * At this point the solver whose sub-pool it is is now removed from
+     * the competition.
+     * This then starts the `delay` timer until exit is allowed.
+     * @param pool The sub-pool to announce an exit for
+     * @dev Emits an event `AnnounceExit`
+     */
+    function announceDestroy(SubPool pool) external onlySubpoolAuth;
+    /**
+     * Withdraw any remaining collateral / cow from the sub-pool.
+     * Funds are sent to `msg.sender`.
+     * @dev Should emit an event `SubPoolDestroyed` on the factory
+     */
+    function finalizeExit(SubPool pool) external onlySubpoolAuth;
+    /**
+     * Set the backend URI for the sub-pool
+     * @param pool The sub-pool to set the URI for
+     * @param newUri The new URI to set
+     * @dev Emits an event `SetBackendUri` on the factory when setting a URI
+     */
+    function setBackendUri(SubPool pool, string calldata newUri) external onlySubpoolAuth;
+}
+
+interface SignedSettlement is Auth {
+    /**
+     * Signed settlement function that is callable by `msg.sender` if `msg.sender` is in
+     * receipt of a valid signature. The signer is a trusted entity and bears the responsibility
+     * of ensuring that `msg.sender` is correctly collateralised.
+     *  
+     * The signed data **MUST** include:
+     * - `solver` - the address of the sub-bonded solver, i.e. `msg.sender`
+     * - Replay protection - an expiry based on some configurable method (e.g. block number)
```

Comment:
This is not necessarily guaranteed as the orders could be partially fillable (and not yet completely filled).

Reply:
Ah, okay, we can add a nonce to the message being signed, but that is going to need an SSTORE post successful execution and an SLOAD pre-execution. An expiry won't necessarily protect against replay, because it can be replayed before the expiry. I think if we want 0 SSTORE/SLOAD this replay violation needs to be dealt offchain with fines probably?

Reply:
> I think if we want 0 SSTORE/SLOAD this replay violation needs to be dealt offchain with fines probably?
This is kind of my thinking where we should have some relatively good guarantees (like the signature is only good for 'x' blocks which gives them an opportunity to land it on chain). 

If it lands on-chain multiple times, then it would be an off-chain fine / slashing event. Otherwise, we hit the SSTORE / SLOAD issue.